gems/smithy-cbor/lib/smithy-cbor/deserializer.rb

```diff
@@ -48,43 +48,56 @@ def map(ref, values, target = nil)
         values.each do |key, value|
           next if value.nil? && !sparse?(ref.shape)
 
-          target[key] = value.nil? ? nil : shape(ref.shape.value, value)
+          target[key] = shape(ref.shape.value, value)
         end
         target
       end
 
       def structure(ref, values, target = nil)
-        return Schema::EmptyStructure.new if ref.shape == Prelude::Unit
-
         target = ref.shape.type.new if target.nil?
         ref.shape.members.each do |member_name, member_ref|
-          key = member_ref.member_name
-          next unless values.key?(key)
-
-          target[member_name] = shape(member_ref, values[key])
+          value = values[member_ref.member_name]
+          value = default(member_ref) if value.nil? && default?(member_ref.traits)
+          target[member_name] = shape(member_ref, value)
         end
         target
       end
 
       def union(ref, values, target = nil) # rubocop:disable Metrics/AbcSize
-        raise ArgumentError, "union value includes more than one key, received: #{values.keys}" if values.size > 1
```

Comment:
Do we have a documentation that confirms that we shouldn't raise when we receive more than one key? I agree that this is probably what we want to do but was curious.

Reply:
It's documented as such in Smithy - we're supposed to select the first non nil value, so we were just doing it wrong to begin with.